### PR TITLE
Improve boot code and Main

### PR DESCRIPTION
**Pull request by eregon**
Let the return value from the TopLevelRaiseHandler flow to Main
* Only use TOP_LEVEL_FIRST for the mains script invocation,
  not for the boot code snippet.

### CODE REVIEW

File: src/main/java/org/truffleruby/core/CoreLibrary.java [hunk 1]
```diff
@@ -1596,7 +1597,7 @@ public Object getTruffleKernelModule() {
             TruffleFixnumNodesFactory.getFactories(),
             TruffleGCNodesFactory.getFactories(),
             TruffleGraalNodesFactory.getFactories(),
-            TruffleKernelNodesFactory.getFactories(),
+                TruffleKernelNodesFactory.getFactories(),
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/core/CoreLibrary.java]: Is this right?

File: src/main/java/org/truffleruby/Main.java [hunk 2]
```diff
@@ -143,11 +143,11 @@ public static void main(String[] args) {
                     if (config.shouldUsePathScript()) {
                         context.setOriginalInputFile(config.getScriptFileName());
                         //language=ruby
-                        bootCode = "Kernel.exit Truffle::Boot.main_s";
+                        bootCode = "Truffle::Boot.main_s";
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/Main.java]: Didn't @nirvdrum just put these in?

File: src/main/java/org/truffleruby/core/CoreLibrary.java (hunk 1 above)
Comment (eregon) [inline src/main/java/org/truffleruby/core/CoreLibrary.java]: Eclipse auto-indent doesn't seem to agree here, I'll fix.

File: src/main/java/org/truffleruby/Main.java (hunk 2 above)
Comment (eregon) [inline src/main/java/org/truffleruby/Main.java]: Only the one for `main_s`. But these `exit` were really just a heavy way to pass the return value to the Main.

Comment (eregon) [inline src/main/java/org/truffleruby/Main.java]: I added a spec for it.

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java [hunk 3]
```diff
@@ -109,13 +109,14 @@ private Object loadMain(IndirectCallNode callNode, String inputFile) {
                     null);
 
             final CodeLoader.DeferredCall deferredCall = getContext().getCodeLoader().prepareExecute(
-                    ParserContext.TOP_LEVEL,
+                    ParserContext.TOP_LEVEL_FIRST,
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: Why was this change necessary?

Comment (eregon) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: It's just for consistency with the line above. Both values have no effect for `prepareExecute()`.
